sql/src/main/java/io/crate/operation/projectors/ProjectingBatchConsumer.java

```diff
@@ -72,17 +72,28 @@ private ProjectingBatchConsumer(BatchConsumer consumer,
                                     RamAccountingContext ramAccountingContext,
                                     ProjectorFactory projectorFactory) {
         this.consumer = consumer;
-        this.projections = projections;
-        this.jobId = jobId;
-        this.ramAccountingContext = ramAccountingContext;
-        this.projectorFactory = projectorFactory;
+        projectors = new ArrayList<>(projections.size());
+
+        boolean projectorsSupportIndependentScrolling = true;
+        for (Projection projection : projections) {
+            Projector projector = projectorFactory.create(projection, ramAccountingContext, jobId);
+            projectors.add(projector);
+
+            if (!projector.providesIndependentScroll()) {
+                projectorsSupportIndependentScrolling = false;
+            }
+        }
+        if (consumer.requiresScroll() && !projectorsSupportIndependentScrolling) {
```

Comment:
`consumer.requiresScroll` could be checked upfront - if it's false it's not necessery to go through all projections at all.